### PR TITLE
dApps staking v3 - part 1

### PR DESCRIPTION
**Pull request by Dinonard**
**Pull Request Summary**

First PR that will introduce part of dApps staking v3 functionality.
Overall implementation will be broken down into multiple PRs to make it easier for reviewers.

**Part 1**
* maintenance mode
* basic operations for registering dApps
* dApp unregistration, reward destination & owner manipulation
* various struct & types to simplify the implementation
* balance lock implementation

After receiving first review, will amend the comments and implement extensive unit tests.

**Check list**
- [x] dApp manipulation & balance lock implemented 
- [x] first review received
- [x] added unit tests

### CODE REVIEW

File: frame/dapp-staking-v3/src/lib.rs
Comment (gitofdeepanshu) [inline frame/dapp-staking-v3/src/lib.rs:360]: what are the scenarios in which contract can't be unregistered?

File: frame/dapp-staking-v3/src/lib.rs
Comment (gitofdeepanshu) [inline frame/dapp-staking-v3/src/lib.rs]: `BlockNumber` is undefined here.

File: frame/dapp-staking-v3/src/lib.rs
Comment (gitofdeepanshu) [inline frame/dapp-staking-v3/src/lib.rs]: is this a typo? `convenient` maybe?

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs:360]: What should be avoided is unregistering a contract while leaving storage items behind, _forever_.
So later on, when PR progresses more, this TODO is a reminder to revisit the issue (I don't have a clear solution for this yet).

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: Right, late minute breaking changes 😂

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: Yes, should be convenience.

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]: I wonder what's the migration plan to v3? If v3 coexists with prev version, probably best to use a different lock ID.

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]:
```suggestion
    Voting {
        #[codec(compact)]
        ends_at: EraNumber,
    }),
```

It is more about personal preference, but I think this is more readable and self-explanatory. `PeriodType` naming includes `Period` space so we don't need to repeat.

If it makes sense, we can apply this to other enum fields.

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]:
```suggestion
/// Rewards pool for lock participants & dApps
```

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]:
```suggestion
        DAppRewardDestinationUpdated {
```

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]:
```suggestion
                    >= T::MinimumLockedAmount::get(),
```

A minimum amout should be valid right?

**Review by shaunxw — COMMENTED** [review]
Looks good in general!

I wonder why we are using `DispatchResultWithPostInfo` as the return type though there is no post info?

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]: Forgot to add this one: do we need `id` when they can be identified via `T::SmartContract`?

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: v3 will replace the old version, and we want to keep the existing locks.

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: I agree 😁 

Haven't gotten to the part of implementation where I use this, but I'd probably get the same idea as you after writing down `PeriodType::VotingPeriod` xD

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: Right, also discovered this via UT 🙂

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: ID was added because of how dApp rewards will be calculated.
Using `T::SmartContract` as identifier works, logically but it takes a lot of space.
To optimize this, a more compact _secundary_ (not sure if that's the right term) is introduced.

Right now it's part of the design (more info in the doc) but it might change in the future.

**Comment by Dinonard** [comment]
> Looks good in general!
> 
> I wonder why we are using `DispatchResultWithPostInfo` as the return type though there is no post info?

Thanks!

For the `DispatchResultWithPostInfo` you're right that it's not needed. I just copied it from the previous pallet for the first extrinsic I added, then reused it afterwards 😁. Fixed it now.

**Comment by Dinonard** [comment]
All functionality intended for this PR has been implemented, and it's ready for final review 🙏

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [reply frame/dapp-staking-v3/src/lib.rs]: I see. Then it makes sense to use the existing lock ID.

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]:
```suggestion
        ExceededMaxNumberOfContracts,
```

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs]:
```suggestion
            Ok(())
```

`into` isn't needed anymore with `DispatchResult` as return type. Same to other calls.

File: frame/dapp-staking-v3/src/test/mock.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/test/mock.rs]: Could we have it executed in `build` function, so it doesn't have to repeat in each UT?

File: frame/dapp-staking-v3/src/lib.rs
Comment (shaunxw) [inline frame/dapp-staking-v3/src/lib.rs:138]: IMO disabled/maintainance mode should also be covered in unit tests.

**Review by shaunxw — COMMENTED** [review]
Looks good overall! Just some small issues.

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs]: Right, I rushed! Will clean it up!

File: frame/dapp-staking-v3/src/test/mock.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/test/mock.rs]: Sure, good suggestion! Nice way to get `before_each` hook 🙂

File: frame/dapp-staking-v3/src/lib.rs
Comment (Dinonard) [reply frame/dapp-staking-v3/src/lib.rs:138]: True, I didn't forget about it though, just thought I'd cover it at the very end.
Personal bias since for `v2`, I added maintenance mode only later.

Will add it now though!

**Review by shaunxw — APPROVED** [review]

**Comment by Dinonard** [comment]
@shaunxw thanks for the approval!

I was thinking, instead of merging it directly, to keep this in a 'live' branch.
All consecutive PRs would be merged to this branch instead (or I could create another branch to aggregate all PRs).

This would keep the review of PRs light(er), while also providing one big overview in the end, before everything is merged.
What do you think?

**Comment by shaunxw** [comment]
Love this idea. It definitely helps with the review process.

**Pull request merged**
